File: 4-1.py
#!/usr/bin/env python

class Passport:
  fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid", "cid"]

  def __init__(self, str):
    for fv in str.split():
      field, value = fv.split(':')
      setattr(self, field, value)

  def valid(self):
    return all(map(lambda f: f == "cid" or getattr(self, f, None),
                   self.fields))

valid = 0
for s in open('4.input').read().split("\n\n"):
  if Passport(s).valid():
    valid += 1

print(valid)

File: 4-1.rb
#!/usr/bin/env ruby

require 'set'

class Passport
  @@fields = [:byr, :iyr, :eyr, :hgt, :hcl, :ecl, :pid, :cid]
  @@fields.each do |field|
    attr_accessor field
  end

  def initialize(str)
    str.split.each do |fv|
      field, value = fv.split(':')
      self.send(field + "=", value)
    end
  end

  def valid?
    @@fields.all? {|f| f == :cid or self.send f}
  end
end

valid = 0
File.read('4.input').split("\n\n").each do |s|
  valid += 1 if Passport.new(s).valid?
end

print("#{valid}\n")
